src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -171,6 +175,27 @@ def prefork(self, options, build_config):
       self._watchman_is_running.wait()
 
     session = self._graph_helper.new_session()
+    if options.for_global_scope().loop:
+      return session, self._prefork_loop(session, options)
+    else:
+      return session, self._prefork_body(session, options)
+
+  def _prefork_loop(self, session, options):
+    iterations = options.for_global_scope().loop_max
+    target_roots = None
+    while iterations and not self.is_killed:
+      try:
+        target_roots = self._prefork_body(session, options)
+      except session.scheduler_session.execution_error_type as e:
+        # Render retryable exceptions raised by the Scheduler.
+        print(e, file=sys.stderr)
+
+      iterations -= 1
+      while iterations and not self.is_killed and not self._loop_condition.wait(timeout=1):
```

Comment:
Can this be restructured to not be nested while loops with control flow, for clarity? I suspect the inner while could probably just become an if...

Reply:
It needs to be a while loop, because `wait` can timeout.